File: client/src/app/services/elo-change.service.ts
import { Injectable } from '@angular/core';
import { Player } from '@app/classes/player';

@Injectable({
    providedIn: 'root',
})
export class EloChangeService {
    changeEloOfPlayers(oldPlayers: Player[]): Player[] {
        const baseEloChangeForFirstOrLast = 20;
        const baseEloChangeForSecondOrThird = 10;
        const eloDisparityFactor = 20;
        const newPlayers: Player[] = [];
        oldPlayers.forEach((val) => newPlayers.push(Object.assign({}, val)));
        const averageElo = this.calculateAverageElo(oldPlayers);
        newPlayers[0].elo += Math.round(baseEloChangeForFirstOrLast + (averageElo - oldPlayers[0].elo) / eloDisparityFactor);
        newPlayers[1].elo += Math.round(baseEloChangeForSecondOrThird + (averageElo - oldPlayers[1].elo) / eloDisparityFactor);
        newPlayers[2].elo -= Math.round(baseEloChangeForSecondOrThird + (averageElo - oldPlayers[2].elo) / eloDisparityFactor);
        newPlayers[3].elo -= Math.round(baseEloChangeForFirstOrLast + (averageElo - oldPlayers[3].elo) / eloDisparityFactor);
        return newPlayers;
    }
    private calculateAverageElo(players: Player[]): number {
        let averageElo = 0;
        for (const player of players) {
            averageElo += player.elo;
        }
        return Math.floor(averageElo / players.length);
    }
}

File: client_leger/lib/services/eloChangeService.dart
import 'package:client_leger/models/player.dart';

class EloChangeService{
  EloChangeService() {}

  List<Player> changeEloOfPlayers(List<Player> oldPlayers ) {
        const baseEloChangeForFirstOrLast = 20;
        const baseEloChangeForSecondOrThird = 10;
        const eloDisparityFactor = 20;
        List<Player> newPlayers = [];
        for (var i = 0;i<oldPlayers.length;i++) {
          newPlayers.add(new Player(oldPlayers[i].name, false));
          newPlayers[i].elo = oldPlayers[i].elo;
        }
        int averageElo = calculateAverageElo(oldPlayers);
        newPlayers[0].elo += (baseEloChangeForFirstOrLast + (averageElo - oldPlayers[0].elo) / eloDisparityFactor).round();
        newPlayers[1].elo += (baseEloChangeForSecondOrThird + (averageElo - oldPlayers[1].elo) / eloDisparityFactor).round();
        newPlayers[2].elo -= (baseEloChangeForSecondOrThird + (averageElo - oldPlayers[2].elo) / eloDisparityFactor).round();
        newPlayers[3].elo -= (baseEloChangeForFirstOrLast + (averageElo - oldPlayers[3].elo) / eloDisparityFactor).round();
        return newPlayers;
    }

    int calculateAverageElo(List<Player> players ) {
        int averageElo = 0;
        for(var player in players) {
          averageElo += player.elo;
        }
        return (averageElo/players.length).floor();
    }
}
